File: crates/python/mirascript/_vm/lib/vm_global/bit.py
from __future__ import annotations
import math

from ...types import VmAny
from .._helpers import _expect_number

__all__ = [
    "b_and",
    "b_or",
    "b_not",
    "b_xor",
    "shl",
    "sar",
    "shr",
]


def _to_int32(x: VmAny, name="x") -> int:
    x = _expect_number(name, x)
    if not math.isfinite(x):
        return 0

    x = int(x)
    return x


def b_and(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")

    result = x & y

    if result & 0x80000000:
        result -= 0x100000000

    return float(result)


def b_or(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")
    result = (x | y) & 0xFFFFFFFF
    if result & 0x80000000:
        result -= 0x100000000
    return float(result)


def b_not(x: VmAny) -> float:
    x = _to_int32(x, "x")
    result = (~x) & 0xFFFFFFFF
    if result & 0x80000000:
        result -= 0x100000000

    return float(result)


def b_xor(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")
    result = (x ^ y) & 0xFFFFFFFF

    if result & 0x80000000:
        result -= 0x100000000

    return float(result)


def shl(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")

    y &= 31  # 只保留低5位（0~31）
    x &= 0xFFFFFFFF  # 保持32位范围
    return ((x << y) | (x >> (32 - y))) & 0xFFFFFFFF


def sar(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")
    result = x >> y

    return float(result)


def shr(x: VmAny, y: VmAny) -> float:
    x = _to_int32(x, "x")
    y = _to_int32(y, "y")
    if x >= 0:
        return float(x >> y)
    else:
        return float((x + 0x100000000) >> y)

File: crates/vm/src/standard_library/global/bit.rs
use crate::standard_library::{global_builtin, number};
use crate::{MiraValue, Runtime};

pub(super) fn to_u32(value: f64) -> u32 {
    if !value.is_finite() || value == 0.0 {
        return 0;
    }
    value.trunc().rem_euclid(4_294_967_296.0) as u32
}

pub(super) fn install(runtime: &mut Runtime) {
    macro_rules! binary {
        ($name: ident, $operation: expr) => {
            global_builtin!(runtime, fn $name(call, args) {
                let left = to_u32(number(call, args, 0, "x")?);
                let right = to_u32(number(call, args, 1, "y")?);
                Ok(MiraValue::number(($operation)(left, right) as f64))
            });
        };
    }
    binary!(b_and, |a: u32, b: u32| (a & b) as i32);
    binary!(b_or, |a: u32, b: u32| (a | b) as i32);
    binary!(b_xor, |a: u32, b: u32| (a ^ b) as i32);
    binary!(shl, |a: u32, b: u32| a.wrapping_shl(b & 31) as i32);
    binary!(sar, |a: u32, b: u32| (a as i32) >> (b & 31));
    binary!(shr, |a: u32, b: u32| a >> (b & 31));
    global_builtin!(runtime, fn b_not(call, args) {
        Ok(MiraValue::number(
            (!to_u32(number(call, args, 0, "x")?) as i32) as f64,
        ))
    });
}
